presto-cache/src/main/java/com/facebook/presto/cache/alluxio/AlluxioCachingClientFileSystem.java

```diff
@@ -0,0 +1,254 @@
+import alluxio.client.file.FileInStream;
+import alluxio.client.file.FileOutStream;
+import alluxio.client.file.URIStatus;
+import alluxio.conf.AlluxioConfiguration;
+import alluxio.grpc.CreateDirectoryPOptions;
+import alluxio.grpc.CreateFilePOptions;
+import alluxio.grpc.DeletePOptions;
+import alluxio.grpc.ExistsPOptions;
+import alluxio.grpc.FreePOptions;
+import alluxio.grpc.GetStatusPOptions;
+import alluxio.grpc.ListStatusPOptions;
+import alluxio.grpc.MountPOptions;
+import alluxio.grpc.OpenFilePOptions;
+import alluxio.grpc.RenamePOptions;
+import alluxio.grpc.ScheduleAsyncPersistencePOptions;
+import alluxio.grpc.SetAclAction;
+import alluxio.grpc.SetAclPOptions;
+import alluxio.grpc.SetAttributePOptions;
+import alluxio.grpc.UnmountPOptions;
+import alluxio.security.authorization.AclEntry;
+import alluxio.security.authorization.Mode;
+import alluxio.wire.BlockLocationInfo;
+import alluxio.wire.FileInfo;
+import alluxio.wire.MountPointInfo;
+import alluxio.wire.SyncPointInfo;
+import com.facebook.presto.hive.HiveFileContext;
+import com.facebook.presto.hive.filesystem.ExtendedFileSystem;
+import org.apache.hadoop.fs.FileStatus;
+import org.apache.hadoop.fs.Path;
+import org.apache.hadoop.fs.permission.FsPermission;
+
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+import static com.google.common.base.Preconditions.checkState;
+import static com.google.common.hash.Hashing.md5;
+import static java.nio.charset.StandardCharsets.UTF_8;
+import static java.util.Objects.requireNonNull;
+
+public class AlluxioCachingClientFileSystem
+        implements alluxio.client.file.FileSystem
+{
+    private final AlluxioConfiguration alluxioConfiguration;
+    private final ExtendedFileSystem fileSystem;
+
+    public AlluxioCachingClientFileSystem(
+            ExtendedFileSystem fileSystem,
+            AlluxioConfiguration alluxioConfiguration)
+    {
+        this.fileSystem = requireNonNull(fileSystem, "fileSystem is null");
+        this.alluxioConfiguration = requireNonNull(alluxioConfiguration, "alluxioConfiguration is null");
+    }
+
+    @Override
+    public void createDirectory(AlluxioURI alluxioURI, CreateDirectoryPOptions options)
+            throws IOException
+    {
+        FsPermission permission = new FsPermission(Mode.fromProto(options.getMode()).toShort());
+        fileSystem.mkdirs(toPath(alluxioURI), permission);
+    }
+
+    @Override
+    public FileInStream openFile(AlluxioURI alluxioURI, OpenFilePOptions options)
+            throws IOException
+    {
+        return new AlluxioCachingInputStream(fileSystem.open(toPath(alluxioURI)));
+    }
+
+    @Override
+    public FileInStream openFile(URIStatus uriStatus, OpenFilePOptions options)
+            throws IOException
+    {
+        checkState(uriStatus instanceof AlluxioURIStatus);
+        HiveFileContext hiveFileContext = ((AlluxioURIStatus) uriStatus).getHiveFileContext();
```

Comment:
Yep, this is one of the most important lines in this PR. Let's make a comment on this. Something like 
```
piggyback hiveFileContext payload through uriStatus
hiveFileContext is critical to openFile interface that cannot be ignored
```